### ISSUE TITLE
Bizarrely large memory increase with batch size in the memory_encoder

### ISSUE THREAD

**Issue by gkioxari** [facebookresearch/segment-anything-2#208]
Hi all!

I am encountering an issue pertaining to the memory of the the HieraDet memory encoder in SAM2. Everything I am reporting runs on an A100 with 80GB.

Consider a SAM2 super dummy model which in pseudocode does the following for an input tensor of `images: (B, 3, 1024, 1024)`
```
def forward(self, images):
    with torch.no_grad():
        backbone_out = self.image_encoder(images)
    preds = linear1(backbone_out["backbone_fpn"][2])
    return preds
```

with a dummy `loss = preds.mean()` and a single linear layer `linear1=nn.Linear(256, 256)`. In this dummy model the only learnable parameters are in the linear layer while the HieraDet memory encoder is kept frozen.

When training this, the memory consumed on the A100 is as follows:
`B=60, mem=50.2GB`
`B=63, mem=52.5GB`
`B>=64, mem=OOM`

This is strange, isn't it?? The strange thing being going from 53G to OOM by going from B=63 to 64. 

My pytorch version is 2.4.0 and CUDA Version: 12.2

**Comment by gkioxari** [comment]
Because I wrap my hiera backbone with `torch.no_grad` during training, my memory used is less when training. As I say in my post, for `B=63`, 52GB is occupied on my A100. For `B=64`, the model goes OOM. Going 52G to OOM when adding one example in the batch seems steep.

**Comment by ronghanghu** [comment]
Hi @gkioxari, thanks for your interest in SAM 2! Regarding the memory consumption of the Hiera backbone, it could sometimes take a lot of memory given is quite large image size (SAM 2 uses a default image size of 1024x1024 following SAM). Especially, going from B=63 to B=64 might change some buffer calculations in Flash Attention or in e.g. PyTorch allocation and reservations.

To reduce memory consumption for your use case, below are possible things that can be tried out:

1. If you haven't done it already, you can put the Hiera backbone forward under bfloat16 automatic mixed precision via `with torch.autocast("cuda", dtype=torch.bfloat16):` context. We are using bfloat16 AMP in our training and inference.
2. Even though the backbone is under `torch.no_grad`, it might still consume some memory. A large part of its allocation might be related to Flash Attention kernel used in the SAM 2 codebase by default. You may switch to the memory-efficient kernel for scaled dot-product by wrapping these lines (https://github.com/facebookresearch/segment-anything-2/blob/0db838b11726893f151fa5826ecfa744e1a7760f/sam2/modeling/backbones/hieradet.py#L68-L72) with a context manager that turns off Flash Attention and uses only memory efficient kernel or math kernel via `with torch.backends.cuda.sdp_kernel(enable_flash=False):` as mentioned in https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html.
3. Setting `export PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True` (as mentioned in https://pytorch.org/docs/stable/notes/cuda.html#using-custom-memory-allocators-for-cuda) might help avoid OOM in some edge cases.
4. In case none of the solutions above works, a last resort is to forward the images in chunks through the Hiera backbone and compute the averages on each chunk.

Hope these would resolve your use case!

**Comment by gkioxari** [comment]
Thanks @ronghanghu! Let me try some of these things and get back to you. I still feel that the steep memory requirements going from B=63 to B64 are bizarre, but oh well! Thank you so much!!
